### PR TITLE
[Design] Intermediate Representation

### PR DESCRIPTION
**Pull request by tonyyang-svail**

### CODE REVIEW

File: doc/design_intermediate_representation.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	DType     string // e.g. "float", "int32"
+	Delimiter string // e.g. ","
+	Shape     []int  // e.g. [1], [1 2 3]
+	IsSparse  bool   // e.g. false
+}
+
+// TrainIR is the intermediate representation for code generation of a training job
+type TrainIR struct {
+	DataSource  string                 // e.g. "hive://root:root@localhost:10000/churn"
+	ExtraConfig string                 // Extra configuration in JSON format. e.g. OSS credential
```
Comment (typhoonzero) [inline doc/design_intermediate_representation.md]: We don't need an `ExtraConfig`, we can't judge what's "extra" and what's not. If it is some configuration, just put in attributes.

File: doc/design_intermediate_representation.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	Select      string                 // e.g. "select * from iris.train"
+	ExtraSelect map[string]string      // e.g. {"validation": "select * from iris.val;"}
```
Comment (typhoonzero) [inline doc/design_intermediate_representation.md]: Same as above, if it's only used to store validation sql, just name it `ValidationSelect`.

File: doc/design_intermediate_representation.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	Estimator   string                 // e.g. "DNNClassifier"
+	Attribute   map[string]interface{} // e.g. {"train.epoch": 1000, "model.hidden_units": [10 10]}
+	Feature     map[string]FieldMeta   // e.g. {"sepal_length": {"float", "", [1], false}, ...}
```
Comment (typhoonzero) [inline doc/design_intermediate_representation.md]: `Feature` only have information about how to parse the column data, we still need information about how to construct a feature column on it.

File: doc/design_intermediate_representation.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	DType     string // e.g. "float", "int32"
+	Delimiter string // e.g. ","
+	Shape     []int  // e.g. [1], [1 2 3]
+	IsSparse  bool   // e.g. false
+}
+
+// TrainIR is the intermediate representation for code generation of a training job
+type TrainIR struct {
+	DataSource  string                 // e.g. "hive://root:root@localhost:10000/churn"
+	ExtraConfig string                 // Extra configuration in JSON format. e.g. OSS credential
```
Comment (weiguoz) [inline doc/design_intermediate_representation.md]: I would suggest using a field name with clearer meaning, or people may be confused when they see it.

File: doc/design_intermediate_representation.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	Select      string                 // e.g. "select * from iris.train"
+	ExtraSelect map[string]string      // e.g. {"validation": "select * from iris.val;"}
+	Estimator   string                 // e.g. "DNNClassifier"
+	Attribute   map[string]interface{} // e.g. {"train.epoch": 1000, "model.hidden_units": [10 10]}
```
Comment (typhoonzero) [inline doc/design_intermediate_representation.md]: https://github.com/sql-machine-learning/sqlflow/issues/774 here's some of my thoughts about refactoring attribute resolution.

**Review by typhoonzero — COMMENTED** [review]
Agree to have IR now, we still need to look into the details of the IR struct.

File: doc/design_intermediate_representation.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+## Intermediate Representation
+
+We propose the following struct as the IR for code generation.
+
+```go
+package sql
+
+// FieldMeta contains the meta information for decoding
+type FieldMeta struct {
+	DType     string // e.g. "float", "int32"
```
Comment (weiguoz) [inline doc/design_intermediate_representation.md]: How about we define several `enmu` types? like [http/status.go](https://golang.org/src/net/http/status.go)

File: doc/design_intermediate_representation.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	Label       map[string]FieldMeta   // e.g. {"class": {"int32", "", [1], false}}
+}
+
+// PredictIR is the intermediate representation for code generation of a prediction job
+type PredictIR struct {
+	DataSource  string                 // e.g. "hive://root:root@localhost:10000/churn"
+	ExtraConfig string                 // Extra configuration in JSON format. e.g. OSS credential
+	Select      string                 // e.g. "select * from iris.train"
+	Estimator   string                 // e.g. "DNNClassifier"
+	Attribute   map[string]interface{} // e.g. {"predict.batch_size": 32}
```
Comment (weiguoz) [inline doc/design_intermediate_representation.md]: Attribute(s)?

File: doc/design_intermediate_representation.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+// FieldMeta contains the meta information for decoding
+type FieldMeta struct {
+	DType     string // e.g. "float", "int32"
+	Delimiter string // e.g. ","
+	Shape     []int  // e.g. [1], [1 2 3]
+	IsSparse  bool   // e.g. false
+}
+
+// TrainIR is the intermediate representation for code generation of a training job
```
Comment (Yancey0623) [inline doc/design_intermediate_representation.md:62]: Maybe we don't need to implement an IR for each job, how about simplifying like: 
``` golang
type FeatureMeta struct {
    DType string
    Delimiter string 
    ...
}
type DBConn struct {
    Driver string
    User string
    ....
}
type ClauseIR struct {
    Estimator string
    SelectClause string
    Attributes map[string]interface{}
    DBConn DBConn
    Features map[string]FeatureMeta
    ...
}
``` 

Each generator can extend the `ClauseIR` as needed.

File: doc/design_intermediate_representation.md (hunk 6 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: Sounds great. `enum` types are more informative than opaque string, so that the developers know what to expect.

File: doc/design_intermediate_representation.md (hunk 4 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: Sure.

File: doc/design_intermediate_representation.md (hunk 1 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: I put `ExtraConfig` here for possible extensibility. I don't have a clear idea of what to put at `ExtraConfig`. So I will remove this field.

File: doc/design_intermediate_representation.md (hunk 2 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: Sure.

File: doc/design_intermediate_representation.md (hunk 7 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: I think singular is fine. The `map` type already indicates there are multiple attributes. The same reasoning applies to `Feature`.

File: doc/design_intermediate_representation.md (hunk 8 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md:62]: @Yancey1989 Thanks for the suggestion. Combining all three `IR`s to a single `ClauseIR` does save some code. However, I still advocate using separate IRs for different job types. Here is my reasoning.
- Avoid confusion. The developer of `xgboost.Predict` would be confused by the `ValidationSelect` field in `ClauseIR`. Also, as we adding more features to SQLFlow, more fields would be added to `CluaseIR`, and the confusion will increase.
- Less work. We either distinguish the job type in `sql` or in `codegen`. However, there are many `codegen`s and only one `sql`. Distinguishing the job type in `sql` saves works in all `codegen`s.

File: doc/design_intermediate_representation.md (hunk 3 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: Thanks. I added `FeatureColumn` field in `FieldMeta`.

File: doc/design_intermediate_representation.md (hunk 5 above)
Comment (tonyyang-svail) [reply doc/design_intermediate_representation.md]: @typhoonzero Thanks for the link. I think `Attribute map[string]interface{}` here is fine.

I believe `AttrMeta` belong to the code generation package since different machine learning toolkits accept different attributes.

Comment (typhoonzero) [reply doc/design_intermediate_representation.md]: > I believe AttrMeta belong to the code generation package since different machine learning toolkits accept different attributes

It is. Each submitter's code generation package define it's `AttrMeta` and call a function to get an `Attribute map[string]interface{}`.

**Review by typhoonzero — APPROVED** [review]
LGTM generally

**Review by weiguoz — APPROVED** [review]
LGTM

**Review by Yancey0623 — APPROVED** [review]
LGTM
